app/controllers/application_controller.rb

```diff
@@ -1,11 +1,54 @@
 class ApplicationController < ActionController::Base
   protect_from_forgery with: :exception
-  before_action :configure_permitted_parameters, if: :devise_controller?
+
+  rescue_from CanCan::AccessDenied do |exception|
+    redirect_to root_url, :alert => exception.message
+  end
+
+  def current_or_guest_user
+    if current_user
+      if session[:guest_user_id] && session[:guest_user_id] != current_user.id
+        # reload guest_user to prevent caching problems before destruction
+        guest_user(with_retry = false).reload.try(:destroy)
+        session[:guest_user_id] = nil
+      end
+      current_user
+    else
+      guest_user
+    end
+  end
+
+  def current_ability
+    @current_ability ||= Ability.new(current_or_guest_user)
+  end
+
+  # find guest_user object associated with the current session,
+  # creating one as needed
+  def guest_user(with_retry = true)
+    # Cache the value the first time it's gotten.
+    @cached_guest_user ||= User.find(session[:guest_user_id] ||= create_guest_user.id)
+  rescue ActiveRecord::RecordNotFound # if session[:guest_user_id] invalid
+    session[:guest_user_id] = nil
+    guest_user if with_retry
+  end
 
   private
 
-  def configure_permitted_parameters
-    devise_parameter_sanitizer.permit(:sign_up, keys: [:name])
-    devise_parameter_sanitizer.permit(:account_update, keys: [:name, :picture])
+  def redirect_to_back_or_default(default = root_url)
+    if request.env["HTTP_REFERER"].present? and request.env["HTTP_REFERER"] != request.env["REQUEST_URI"]
+      redirect_to :back
+    else
+      redirect_to default
+    end
   end
+
+  def create_guest_user
+    u = User.create(name: "Guest",
+                    email: "guest_#{Time.now.to_i}#{rand(100)}@example.com",
+                    guest: true)
+    u.save!(:validate => false)
```

Comment:
Also, this is a good case for introducing you to builder objects. :)

Pieces of logic such as this one (building some object in a specific way) don't belong in a controller. Controllers are for params sanitization, enforcing constraints (user auth, etc.), redirects. The actual logic should be elsewhere. Extract this code to a builder class:

``` ruby
# app/builders/user_builders/guest.rb
    module UserBuilders
      class Guest
        def build
          User.new(
            name:  "Guest",
            email: generate_guest_email,
            guest: true,
          )
        end

        private

        def generate_guest_email
          "guest_#{Time.now.to_i}#{rand(100)}@example.com"
        end
      end
    end
```
Then 

``` ruby
guest = UserBuilders::Guest.new.build.save(validate: false)
session[:guest_user_id] = guest.id
guest
```